libs/openchallenges/ui/src/lib/challenge-card/challenge-card.component.ts

```diff
@@ -63,6 +57,48 @@ export class ChallengeCardComponent implements OnInit {
         : this.imageService.getImage({
             objectKey: 'banner-default.svg',
           });
+      if (this.challenge.endDate && this.status === 'completed') {
+        const timeSince = this.calcTimeDiff(this.challenge.endDate);
+        if (timeSince) {
+          this.time_info = `Ended ${timeSince} ago`;
+        }
+      } else if (this.challenge.endDate && this.status === 'active') {
+        this.time_info = `Ends in ${this.calcTimeDiff(this.challenge.endDate)}`;
+      } else if (this.challenge.startDate && this.status === 'upcoming') {
+        this.time_info = `Starts in ${this.calcTimeDiff(
+          this.challenge.startDate
+        )}`;
+      }
     }
   }
+
+  calcTimeDiff(date: string) {
+    const pattern = /\d{4}-\d{2}-\d{2}/;
+    if (!pattern.test(date)) {
+      return '';
+    }
```

Comment:
This function `calcTimeDiff` should validate the input data as you do but validation error should be handled instead of silently returning an empty string.

One approach is to `throw Error(`The input date does not match the schema ${pattern}: ${date}`)`. Then try/catch the function where it is called. The catch block should at least log the error with `console.error()` (note to myself: we should start using a logger within the app).

In the catch block you could then set `timeSince` to `undefined`, which is preferable to an empty string, since an empty since is probably not an expected value for `timeSince`.

Note that a function that may throw an error also has the return type `never`. The return type of the function would then be `string | never`.

Another approach is to change the function `calcTimeDiff` to require a `Date` object as input. The validation would happen where the Date object is constructed. The validation code in `calcTimeDiff` could be removed and the return type would simply be `string`.

The advantage of the second solution is that the function could be reused directly with "native" `Date` objects without the need to do `Date` => `string` => `Date`.

Reply:
To be addressed in a different PR.